packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -968,6 +995,96 @@ export class SnapController extends BaseController<
     );
   }
 
+  #handlePreinstalledSnaps(preinstalledSnaps: PreinstalledSnap[]) {
+    for (const { snapId, manifest, files, removable } of preinstalledSnaps) {
+      const existingSnap = this.get(snapId);
+      const isAlreadyInstalled = existingSnap !== undefined;
+      const isUpdate =
+        isAlreadyInstalled && gtVersion(manifest.version, existingSnap.version);
+
+      // Disallow downgrades and overwriting non preinstalled snaps
+      if (
+        isAlreadyInstalled &&
+        (!isUpdate || existingSnap.preinstalled !== true)
+      ) {
+        continue;
+      }
+
+      const manifestFile = new VirtualFile<SnapManifest>({
+        path: NpmSnapFileNames.Manifest,
+        value: JSON.stringify(manifest),
+        result: manifest,
+      });
+
+      const virtualFiles = files.map(
+        ({ path, value }) => new VirtualFile({ value, path }),
+      );
+      const { filePath, iconPath } = manifest.source.location.npm;
+      const sourceCode = virtualFiles.find((file) => file.path === filePath);
+      const svgIcon = iconPath
+        ? virtualFiles.find((file) => file.path === iconPath)
+        : undefined;
+
+      assert(sourceCode, 'Source code not provided for preinstalled snap');
+
+      assert(
+        manifest.source.files === undefined,
+        'Auxiliary files are not currently supported for preinstalled snaps',
+      );
+
+      assert(
+        manifest.source.locales === undefined,
+        'Localization files are not currently supported for preinstalled snaps',
+      );
+
+      const filesObject: FetchedSnapFiles = {
+        manifest: manifestFile,
+        sourceCode,
+        svgIcon,
+        auxiliaryFiles: [],
+        localizationFiles: [],
+      };
+
+      // Add snap to the SnapController state
+      this.#set({
+        id: snapId,
+        origin: 'MetaMask',
+        files: filesObject,
+        removable,
+        preinstalled: true,
+      });
+
+      // Setup permissions
+      const processedPermissions = processSnapPermissions(
+        manifest.initialPermissions,
+      );
+
+      this.#validateSnapPermissions(processedPermissions);
+
+      const { newPermissions, unusedPermissions } =
+        this.#calculatePermissionsChange(snapId, processedPermissions);
+
+      const unusedPermissionsKeys = Object.keys(unusedPermissions);
+      if (isNonEmptyArray(unusedPermissionsKeys)) {
+        this.messagingSystem.call('PermissionController:revokePermissions', {
+          [snapId]: unusedPermissionsKeys,
+        });
+      }
+
+      if (isNonEmptyArray(Object.keys(newPermissions))) {
+        this.messagingSystem.call('PermissionController:grantPermissions', {
+          approvedPermissions: newPermissions,
+          subject: { origin: snapId },
+        });
+      }
+
+      // Set status
+      this.update((state) => {
+        state.snaps[snapId].status = SnapStatus.Stopped;
+      });
```

Comment:
Can this logic be shared with the existing update logic?

Reply:
Yeah probably!